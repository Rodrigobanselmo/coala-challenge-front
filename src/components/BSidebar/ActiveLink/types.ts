import { ReactElement } from "react";

import { LinkProps } from "next/link";

export interface IActiveLinkProps extends Omit<LinkProps, "href" | "onClick"> {
  children: ReactElement;
  href?: string;
  canOpen?: boolean;
  isOpen?: boolean;
}
