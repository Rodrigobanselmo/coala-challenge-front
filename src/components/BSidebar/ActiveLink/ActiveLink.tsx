import { cloneElement } from "react";

import { Box } from "@mui/material";
import Link from "next/link";
import { useRouter } from "next/router";

import { IActiveLinkProps } from "./types";

export function ActiveLink({
  children,
  ...rest
}: IActiveLinkProps): JSX.Element {
  const { asPath } = useRouter();

  const isBasePath = asPath == "/";

  const getIsActive = () => {
    if (isBasePath) {
      if (rest.href == "/") return true;
    } else {
      if (rest.href == "/") return false;

      return (
        asPath.startsWith(String(rest.href)) ||
        asPath.startsWith(String(rest.href))
      );
    }
  };

  return (
    <Link href={rest.href || ""} {...rest}>
      {cloneElement<any>(children, {
        is_active: getIsActive() ? 1 : 0,
      })}
    </Link>
  );
}
