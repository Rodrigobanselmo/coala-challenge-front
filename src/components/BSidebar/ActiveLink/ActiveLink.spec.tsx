import React from "react";
import { render, screen } from "@testing-library/react";
import { ActiveLink } from "./ActiveLink";

jest.mock("next/router", () => {
  let currentAsPath = "/";
  return {
    useRouter: () => ({
      asPath: currentAsPath,
    }),
    setAsPath: (newAsPath: string) => {
      currentAsPath = newAsPath;
    },
  };
});

describe("ActiveLink", () => {
  beforeEach(() => {
    require("next/router").setAsPath("/");
  });

  it("should render with active status when href matches asPath", () => {
    const { asFragment } = render(
      <ActiveLink href="/">
        <p>Home</p>
      </ActiveLink>
    );

    const myElem = screen.getByText("Home");

    expect(asFragment()).toMatchSnapshot();
    expect(myElem).toHaveAttribute("is_active", "1");
  });

  it("should render without active status when href does not match asPath", () => {
    const { asFragment } = render(
      <ActiveLink href="/about">
        <p>Home</p>
      </ActiveLink>
    );

    const myElem = screen.getByText("Home");

    expect(asFragment()).toMatchSnapshot();
    expect(myElem).toHaveAttribute("is_active", "0");
  });

  it("should render with a equal asPath value", () => {
    // Set a different asPath value for this test
    require("next/router").setAsPath("/some-other-path");

    const { asFragment } = render(
      <ActiveLink href="/some-other-path">
        <p>Some Other Path</p>
      </ActiveLink>
    );

    const myElem = screen.getByText("Some Other Path");

    expect(asFragment()).toMatchSnapshot();
    expect(myElem).toHaveAttribute("is_active", "1");
  });

  it("should render with a different asPath value", () => {
    require("next/router").setAsPath("/other-path-2");

    const { asFragment } = render(
      <ActiveLink href="/some-other-path">
        <p>Some Other Path</p>
      </ActiveLink>
    );

    const myElem = screen.getByText("Some Other Path");

    expect(asFragment()).toMatchSnapshot();
    expect(myElem).toHaveAttribute("is_active", "0");
  });
});
