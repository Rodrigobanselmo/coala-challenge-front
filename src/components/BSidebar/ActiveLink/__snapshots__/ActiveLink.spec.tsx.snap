// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveLink should render with a different asPath value 1`] = `
<DocumentFragment>
  <a
    href="/some-other-path"
  >
    <p
      is_active="0"
    >
      Some Other Path
    </p>
  </a>
</DocumentFragment>
`;

exports[`ActiveLink should render with a equal asPath value 1`] = `
<DocumentFragment>
  <a
    href="/some-other-path"
  >
    <p
      is_active="1"
    >
      Some Other Path
    </p>
  </a>
</DocumentFragment>
`;

exports[`ActiveLink should render with active status when href matches asPath 1`] = `
<DocumentFragment>
  <a
    href="/"
  >
    <p
      is_active="1"
    >
      Home
    </p>
  </a>
</DocumentFragment>
`;

exports[`ActiveLink should render without active status when href does not match asPath 1`] = `
<DocumentFragment>
  <a
    href="/about"
  >
    <p
      is_active="0"
    >
      Home
    </p>
  </a>
</DocumentFragment>
`;
